store/streaming/file/service.go

```diff
@@ -20,262 +23,134 @@ var _ baseapp.StreamingService = &StreamingService{}
-	if err != nil {
-		return err
-	}
-	if _, err = dstFile.Write(lengthPrefixedReqBytes); err != nil {
-		return err
-	}
-	// write all state changes cached for this stage to file
-	fss.stateCacheLock.Lock()
-	for _, stateChange := range fss.stateCache {
-		if _, err = dstFile.Write(stateChange); err != nil {
-			fss.stateCache = nil
-			fss.stateCacheLock.Unlock()
-			return err
-		}
-	}
-	// reset cache
-	fss.stateCache = nil
-	fss.stateCacheLock.Unlock()
-	// write res to file
-	lengthPrefixedResBytes, err := fss.codec.MarshalLengthPrefixed(&res)
-	if err != nil {
-		return err
-	}
-	_, err = dstFile.Write(lengthPrefixedResBytes)
-	return err
-}
-
-func (fss *StreamingService) openDeliverTxFile() (*os.File, error) {
-	fileName := fmt.Sprintf("block-%d-tx-%d", fss.currentBlockNumber, fss.currentTxIndex)
-	if fss.filePrefix != "" {
-		fileName = fmt.Sprintf("%s-%s", fss.filePrefix, fileName)
-	}
-	fss.currentTxIndex++
-	return os.OpenFile(filepath.Join(fss.writeDir, fileName), os.O_CREATE|os.O_WRONLY, 0o600)
+	fss.blockMetadata.DeliverTxs = append(fss.blockMetadata.DeliverTxs, &types.BlockMetadata_DeliverTx{
+		Request:  &req,
+		Response: &res,
+	})
+	return nil
 }
 
 // ListenEndBlock satisfies the baseapp.ABCIListener interface
 // It writes the received EndBlock request and response and the resulting state changes
 // out to a file as described in the above the naming schema
 func (fss *StreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) (rerr error) {
-	// generate the new file
-	dstFile, err := fss.openEndBlockFile()
-	if err != nil {
-		return err
-	}
-	defer func() {
-		cerr := dstFile.Close()
-		if rerr == nil {
-			rerr = cerr
-		}
-	}()
+	fss.blockMetadata.RequestEndBlock = &req
+	fss.blockMetadata.ResponseEndBlock = &res
+	return nil
+}
 
-	// write req to file
-	lengthPrefixedReqBytes, err := fss.codec.MarshalLengthPrefixed(&req)
-	if err != nil {
-		return err
-	}
-	if _, err = dstFile.Write(lengthPrefixedReqBytes); err != nil {
-		return err
+func (fss *StreamingService) ListenCommit(ctx sdk.Context, res abci.ResponseCommit) error {
+	fss.blockMetadata.ResponseCommit = &res
+
+	// write to target file atomically
+	metaFileName := fmt.Sprintf("block-%d-meta", fss.currentBlockNumber)
+	dataFileName := fmt.Sprintf("block-%d-data", fss.currentBlockNumber)
+	if fss.filePrefix != "" {
+		metaFileName = fmt.Sprintf("%s-%s", fss.filePrefix, metaFileName)
+		dataFileName = fmt.Sprintf("%s-%s", fss.filePrefix, dataFileName)
 	}
-	// write all state changes cached for this stage to file
-	fss.stateCacheLock.Lock()
-	for _, stateChange := range fss.stateCache {
-		if _, err = dstFile.Write(stateChange); err != nil {
-			fss.stateCache = nil
-			fss.stateCacheLock.Unlock()
+
+	if fss.outputMetadata {
+		bz, err := fss.codec.Marshal(&fss.blockMetadata)
+		if err != nil {
+			return err
+		}
+		if err := atomic.WriteFile(path.Join(fss.writeDir, metaFileName), bytes.NewReader(bz)); err != nil {
```

Comment:
No reason to bring in a new dependency for this, os.WriteFile is sufficient.

Reply:
If the files are to be read by different process, atomicity is necessary to prevent reading partial file.
For example, we might use the file streamer output to replicate the state to distributed nodes, so the atomicity is needed here.

Reply:
> 1. There is no documented guarantee that files produced by this type are written atomically. It cannot be assumed by consumers.

I can change the ADR to add this.

> 2. Regardless, that property is true even without the atomic dependency: this code won't produce a partial file, since the entire content is written in a single call to Write.
> 3. The content is likely only readable to out of process consumers after a sync and/or close, anyway (depending on OS/FS).

I think there's no atomicity guarantee even with a single sync/close call, I guess it depends on the block level implementation.

> 4. Replication of a file on disk to another host doesn't rely on the file being written atomically.

I guess there are ways for the consumer to workaround the possible of partial written file, but atomicity certainly make the consumer's life much easier.

Reply:
There's no reason for it.

Reply:
> There's no reason for it.

do you mean no reason for atomic writing? how should a concurrent consumer read the files reliably?

Reply:
it’s 2 step, if we panic [here](https://github.com/golang/go/blob/master/src/os/file.go#L738) then file is created/truncated but not written, while atomic WriteFile is create tmp file then rename it to original, so no intermediate state

Reply:
Another way to cope is always write the length of the whole content first, so the reader can detect a incomplete file.

Reply:
> do you mean no reason for atomic writing?

Yes. os.WriteFile provides sufficient atomicity for most use cases, including this one.

> how should a concurrent consumer read the files reliably without it?

In general, os.WriteFile does not produce output to the file system that can be read by consumers in an incomplete state. Even in the cases where it does, the file is not synced until the complete content is written, and that syncing operation defines the EOF, so to speak. 

There's no peactical situation where writing a file using normal stdlib functions will produce a normal readable file on disk that only contains some of the written bytes.

Reply:
> if we panic here

io.Writer implementations don't panic. And even if they did, panics aren't meant to be things your code should detect or defend against.

Reply:
> > if we panic here
> 
> io.Writer implementations don't panic. And even if they did, panics aren't meant to be things your code should detect or defend against.

yeah, it's not about panic, mainly about concurrent reader.

> In general, os.WriteFile does not produce output to the file system that can be read by consumers in an incomplete state.

there are several cases here:
- open and write are two operations which are not atomic, but it's easy to workaround by the reader.
- multiple writes are not atomic, which is the case of [current writing code](https://github.com/cosmos/cosmos-sdk/blob/main/store/streaming/file/service.go#L169), so if we want to remove atomic library, at least we need to write to buffer and flush once.
- the last issue, is a single write operation atomic? I just did some research, I think the answer varies for different filesystems, for example, [ext4 seems to not be atomic](https://stackoverflow.com/a/10651090), and if OS don't guarantee atomicity, stdlib can't do.

Reply:
Each file write operation is, as far as I can tell, for a new and unique file. That is, files are never modified, they are only created. In order for consumers to read these files in a partially written state, this would have to happen:

1. ListenCommit makes a single call to io.WriteFile with a single []byte
2. io.WriteFile partially writes the file to disk and is interrupted
3. Consumer(s) are somehow alerted that this file descriptor now exists, and attempt a read operation
4. Those consumer(s) open the file descriptor, do the read, and believe the partial data is the complete file
5. io.WriteFile completes the rest of its write, and closes the file

To the best of my knowledge, there is no realistic situation which would allow step 3 or step 4 to happen. Consumers who want to read the file would realistically only be notified when the writer closes the file. Even if they opened it in an intermediate state, any practical API that reads file descriptors would make it clear that the file is in an intermediate state.

I'm happy to be proven wrong! If you can link me a reproducible test, or even links to docs, I'd love to see them.

Reply:
First of all, when the file is opened, it's already observable as an empty file by the reader.

About if the reader can observe partial write, here is the standard spec: 
https://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html

> After a write() to a regular file has successfully returned:
> 
> Any successful read() from each byte position in the file that was modified by that write shall return the data specified by the write() for that position until such byte positions are again modified.

It only talks about byte level guarantees.
The write is also allowed to write partial content when some system limit is reached.
I might do the test on the machine I work with, but I'm not sure how much assurance it provides even if it success on one machine.

Reply:
[`os.WriteFile`](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/os/file.go;l=721) simply do `OpenFile` followed by `f.Write` without locking, so it's definitely not atomic as a whole.
It treats the short write as an error, but it won't revert the write, in that case, the file is left partially written.

Reply:
but the atomic library do a fsync, which could have a visible performance impact, so let me write the size of the file to the begining of the file, and do normal file writing, so at least consumer can detect a corrupted file.

Reply:
Previously consumers could do e.g.

```go
buf, ... := os.ReadFile(filename)
val, ... := whatever.ParseBytes(buf)
```

Which would have certainly errored out during ParseBytes, if the file was only partially written. Now, consumers must do e.g.

```go
buf, ... := os.ReadFile(filename)
u, dat := buf[:8, 8:]
len, ... := (parse uint64 or error)
val, ... := whatever.ParseBytes(dat)
```

Where is this uint64-length-prefixed file format documented? What affordances are you providing consumers to read values from these files? Are you expecting them to do this work manually? How should they do that work if they aren't consuming the file from a Go (or other) program?

Reply:
> Which, for the record, will certainly error out during ParseBytes in the case that the file is only partially written.

There are many ways that a partial file is still a valid protobuf message, when truncated at field boundary, and there's the data file which is a stream of kvpair messages, when truncated at message boundary, it's still a valid file. this is the exact issue the prefixed length trying to solve.

> How should they do that work if they aren't consuming the file from a Go (or other) program?

It's just a big endian encoded 8 bytes integer, should be easy to parse from any language?

> Where is this uint64-length-prefixed file format documented?

good catch, I'll update the file streamer's `README.md` to reflect the new changes.